rero_ils/modules/loans/cli.py

```diff
@@ -38,12 +38,145 @@
 from ..notifications.api import Notification
 from ..notifications.tasks import create_notifications
 from ..patron_transaction_events.api import PatronTransactionEvent
+from ..patron_transactions.api import PatronTransaction
 from ..patron_types.api import PatronType
 from ..patrons.api import Patron, PatronsSearch
 from ..utils import extracted_data_from_ref, get_ref_for_pid, \
     get_schema_for_resource
 
 
+def check_missing_fields(transaction, transaction_type):
+    """Return list of missing fields for a given transaction type.
+
+    transaction: the json transaction record.
+    transaction_type: type of transaction.
+    """
+    if transaction_type == 'checkout':
+        fields = ['item_pid', 'patron_pid', 'end_date', 'transaction_date',
+                  'transaction_location_pid', 'transaction_user_pid',
+                  'organisation', 'start_date']
+    elif transaction_type == 'request':
+        fields = ['item_pid', 'patron_pid', 'organisation', 'transaction_date',
+                  'pickup_location_pid', 'transaction_location_pid',
+                  'transaction_user_pid', 'request_expire_date']
+    elif transaction_type == 'fine':
+        fields = ['note', 'type', 'patron', 'status', 'organisation',
+                  'total_amount', 'creation_date']
+
+    return [field for field in fields if field not in transaction]
+
+
+def build_loan_record(transaction, transaction_type, item):
+    """Build the loan record before inserting into db.
+
+    transaction: the json transaction record.
+    transaction_type: type of transaction.
+    item: the item record.
+    """
+    if transaction_type == 'checkout':
+        transaction['state'] = 'ITEM_ON_LOAN'
+        transaction['trigger'] = 'checkout'
+        transaction['document_pid'] = item.document_pid
+        transaction['to_anonymize'] = False
+        transaction['item_pid'] = {'value': transaction.get('item_pid'),
+                                   'type': 'item'}
+    elif transaction_type == 'request':
+        transaction['state'] = 'PENDING'
+        transaction['trigger'] = 'request'
+        transaction['item_pid'] = {'value': transaction.get('item_pid'),
+                                   'type': 'item'}
+        transaction['document_pid'] = item.document_pid
+        transaction['to_anonymize'] = False
+
+
+@click.command('load_virtua_transactions')
+@click.option('-t', '--transaction_type', 'transaction_type', is_flag=False,
+              default='checkouts')
+@click.option('-v', '--verbose', 'verbose', is_flag=True, default=False)
+@click.option('-d', '--debug', 'debug', is_flag=True, default=False)
+@click.argument('infile', type=click.File('r'))
+@with_appcontext
+def load_virtua_transactions(infile, transaction_type, verbose, debug):
+    """Load Virtua circulation transactions.
+
+    infile: Json Virtua transactions file
+    transaction_type: Transaction type either checkout, request or fine.
+    """
+    file_data = json.load(infile)
```

Comment:
If the input files could be big it would be preferable to implement lazy loading.

Reply:
Save of problematic records would be nice.